### PR TITLE
MicroUI Flush Visualizer doc

### PR DESCRIPTION
**Pull request by gjadi-mej**
https://docs.microej.com/en/feature-m0269md-243_microuiflushvisualizer/ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.html

### CODE REVIEW

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+1. *Layout*: This phase determines which widgets should be displayed on the screen and their respective positions.
+   It is typically triggered when widgets are added or removed from the widget hierarchy.
+   To ensure efficiency, an application should only modify the widget hierarchy when necessary and avoid doing so during animations.
+
+2. *Rendering*: During this phase, each widget executes its rendering code to perform the necessary drawing operations.
+   It is crucial for widgets to render only what is necessary and minimize overlapping with other widgets to ensure optimal performance.
+
+3. *Flush*: This phase involves copying the UI working buffer to the screen buffer.
+   This operation is performed by the VEE Port, and it is the responsibility of the Firmware developer to optimize this process, for example, by utilizing DMA2D.
```
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:32]: to not mention the DMA2D (not compliant with NXP users)

```suggestion
   This operation is performed by the VEE Port, and it is the responsibility of the Firmware developer to optimize this process, for example, by utilizing a GPU.

```

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+The MicroUI Flush Visualizer is a tool designed to visualize the drawing operations requested by widgets during the *Rendering* phase between each flush.
+
+
+The MicroUIUI Flush Visualizer provides the following information:
```
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:37]:
```suggestion
The MicroUI Flush Visualizer provides the following information:

```

File: ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png:1]: remove this image (not compatible with arch 8 / sdk 5.7)

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 3]
```diff
@@ -0,0 +1,70 @@
+.. _microuiflushvisualizer:
+
+MicroUI Flush Visualizer
+==============================================================================================================
+Visualize the pixel surface drawn between 2 flushes to identify performance bottlenecks in your UI application
```
[the remark this answers is not in the document]
Comment (stephanemainchain) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:5]: I would rephrase into:
Visualize the pixel surface drawn between 2x MicroUI frame buffer flushes to identify performance bottlenecks in your UI application.

File: ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png:1]: Done.

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+MicroUI Flush Visualizer
+==============================================================================================================
+Visualize the pixel surface drawn between 2 flushes to identify performance bottlenecks in your UI application
+--------------------------------------------------------------------------------------------------------------
+
+Building smooth and visually appealing UI applications require a keen focus on performance.
+To achieve efficient UI rendering, minimizing unnecessary work that consumes valuable CPU time is essential.
+This article introduces the MicroUI Flush Visualizer, a tool designed to investigate potential bottlenecks in UI applications running on the Simulator.
+
+- Requirements: VEE Port with MICROUI 3.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: done.

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+MicroUI Flush Visualizer
+========================
+
+Building smooth and visually appealing UI applications require a keen focus on performance.
+To achieve efficient UI rendering, minimizing unnecessary work that consumes valuable CPU time is essential.
+
+This article introduces the MicroUI Flush Visualizer, a tool designed to investigate potential performance bottlenecks in UI applications running on the Simulator by showing the pixel surface drawn between 2 MicroUI frame buffer flushes.
+
+- Requirements: VEE Port with Pack UI 13.0.0 or higher.
```
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:11]:
```suggestion
- Requirements: VEE Port with UI Pack 13.0.0 or higher.

```

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+- A list of drawing operations that were performed during the *Rendering* phase.
+- The total area covered by the sum of the area drawn by the drawing operations as percentage.
+  A value of 100% indicates that the entire display was drawn once; a value of 200% indicates that the entire display was drawn twice; etc.
+
+Here are examples of the MicroUI Flush Visualizer in action:
+
++----------------------------------------------------------------+--------------------------------------------------------------------+
+|Simple GFX                                                      | Demo Wearable                                                      |
++================================================================+====================================================================+
+| .. image:: images/small-demo-ui-flush-visualizer-simpleGFX.png | .. image:: images/small-demo-ui-flush-visualizer-demo-wearable.png |
```
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:51]: use images with filter & background (board)

File: ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/images/ui-flush-visualizer-fp-launcher.png:1]: remove the file

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+A perfect application has 100% of its display area drawn.
+In practice, a total area drawn between 100% to 200% is the norm because widgets often overlap.
+However, if the total area drawn is bigger than 200%, that means that the total surface of the display was drawn more than twice, meaning that a lot of time could be spent drawing things that are never shown.
+
+As always, when conducting a performance study, measure.
+Use :ref:`systemview` to identify the bottlenecks in your application on the embedded target.
+A total area drawn over 200% is inefficient, but your application may have bigger bottlenecks.
+Confirm it by measuring the time spent drawing vs. the time spent elsewhere between flushes.
+
```
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:64]: you mention the git project at the beginning but when we read the doc, we expect something else at the end. Maybe you can add a final note "For more details, see xxx" ?

File: ApplicationDeveloperGuide/UI/Advanced/images/small-demo-ui-flush-visualizer-demo-wearable.png
Comment (gbalan-microej) [inline ApplicationDeveloperGuide/UI/Advanced/images/small-demo-ui-flush-visualizer-demo-wearable.png:1]: after reflection, I am not sure if it is the best example: you can see the trick that consists to use some images instead of drawing the text!

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+The MicroUI Flush Visualizer is a tool designed to visualize the drawing operations requested by widgets during the *Rendering* phase between each flush.
+
+
+The MicroUI Flush Visualizer provides the following information:
+
+
+- A screenshot of the result of the Rendering Pipeline (what was shown on the screen).
+- A list of drawing operations that were performed during the *Rendering* phase.
+- The total area covered by the sum of the area drawn by the drawing operations as percentage.
+  A value of 100% indicates that the entire display was drawn once; a value of 200% indicates that the entire display was drawn twice; etc.
```
Comment (seonmej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:51]: I don't know exactly how to express it but, the display is not mandatorily drawn several times. But the surface drawn is equivalent to X times the surface of the display.
```suggestion
  A value of 100% indicates that the entire display was drawn once; a value of 200% indicates that the surface of the display was drawn twice; etc.

```

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+Use :ref:`systemview` to identify the bottlenecks in your application on the embedded target.
+A total area drawn over 200% is inefficient, but your application may have bigger bottlenecks.
+Confirm it by measuring the time spent drawing vs. the time spent elsewhere between flushes.
+
+Installation
+------------
+
+1. Open the Front Panel file ``.fp``.
+
+2. Replace the ``ej.fp.widget.Display`` widget by ``ej.fp.widget.FlushVisualizerDisplay``
```
Comment (seonmej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:71]:
```suggestion
2. Replace the ``ej.fp.widget.Display`` widget by ``ej.fp.widget.FlushVisualizerDisplay``.

```

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+3. Set the shape of the display to either ``RECTANGLE`` or ``CIRCLE``:
```
Comment (seonmej) [inline ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst:73]: Unnecessary after discussions on the code.

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+2. Open the file ``MicroUIFlushVisualizer/MicroUIFlushVisualizer.html`` that was generated in the :ref:`application output folder <outputfiles>`.
+
+.. image:: images/MicroUIFlushVisualizerApplicationOutputFolder.png
+
+
+Known Limitations
+-----------------
+
+- Do not take into account the clip of the GraphicsContext.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: done

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst (hunk 10 above)
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: removed

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst (hunk 8 above)
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: I've reformulated this sentence. What do you think?

Comment (seonmej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: Nice!

**Review by seonmej — APPROVED** [review]

**Comment by gjadi-mej** [comment]
Awaiting widgets publication before the merge.

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 12]
```diff
@@ -0,0 +1,95 @@
+.. _microuiflushvisualizer:
+
+MicroUI Flush Visualizer
+========================
+
+Building smooth and visually appealing UI applications require a keen focus on performance.
+To achieve efficient UI rendering, minimizing unnecessary work that consumes valuable CPU time is essential.
+
+This article introduces the MicroUI Flush Visualizer, a tool designed to investigate potential performance bottlenecks in UI applications running on the Simulator by showing the pixel surface drawn between 2 MicroUI frame buffer flushes.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: fixed.

File: ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst [hunk 13]
```diff
@@ -0,0 +1,70 @@
+.. _microuiflushvisualizer:
+
+MicroUI Flush Visualizer
+==============================================================================================================
+Visualize the pixel surface drawn between 2 flushes to identify performance bottlenecks in your UI application
+--------------------------------------------------------------------------------------------------------------
+
+Building smooth and visually appealing UI applications require a keen focus on performance.
```
[the remark this answers is not in the document]
Comment (gjadi-mej) [reply ApplicationDeveloperGuide/UI/Advanced/how-to-microui-flush-visualizer.rst]: Got it, fixed. thank you.
